File: lib/github.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compare = exports.createGitHubClient = void 0;
const rest_1 = require("@octokit/rest");
function createGitHubClient(token) {
    const client = new rest_1.Octokit({
        auth: token,
    });
    return client;
}
exports.createGitHubClient = createGitHubClient;
async function compare(client, payload) {
    const result = await client.rest.repos.compareCommits({
        owner: payload.owner,
        repo: payload.repo,
        base: payload.target,
        head: "main",
    });
    return result;
}
exports.compare = compare;

File: src/github.rs
use octocrab::Octocrab;
use serde::Deserialize;

pub struct GitHub {
    client: Octocrab,
}

#[derive(Deserialize)]
pub struct Author {
    name: String
}


#[derive(Deserialize)]
pub struct Commit {
    author: Author,
    message: String,
}

impl GitHub {
    pub fn new(token: String) -> GitHub {
        match Octocrab::builder().personal_token(token).build() {
            Ok(client) => GitHub { client },
            Err(e) => panic!(e),
        }
    }
}

pub struct ComparePayload {
    pub owner: String,
    pub repo: String,
    pub target: String,
}

#[derive(Deserialize)]
struct CompareResponseCommit {
    commit: Commit
}

#[derive(Deserialize)]
struct CompareResponse {
    commits: Vec<CompareResponseCommit>,
}
pub async fn compare(github: GitHub, payload: ComparePayload) {
    let request_url = format!("https://api.github.com/repos/{}/{}/compare/main...v2.2.0", payload.owner, payload.repo);

    if let Ok(result) = github.client._get(request_url, None::<&()>).await {
        if let Ok(response) = result.json::<CompareResponse>().await {
            for commit in response.commits {
                println!("{}", commit.commit.message);
            }
        }
    }
}
